### PR TITLE
Verify individual messages via cross-signing

### PR DESCRIPTION
**Pull request by foldleft**
Fixes https://github.com/vector-im/riot-web/issues/11880

**user test4 logs out, then back in**
Screenshot 2020-01-20 at 17 25 31

**Deleted devices get warnings**
Screenshot 2020-01-20 at 17 25 37

### CODE REVIEW

**Comment by jryans** [comment]
@foldleft It looks you may have added new strings, in which case you'll need to run `yarn i18n` to update the translation file and commit the result.

File: src/components/views/rooms/EventTile.js [hunk 1]
```diff
@@ -282,18 +284,42 @@ export default createReactClass({
...
+            });
+            return;
+        }
+
+        const eventSenderTrust = await this.context.checkEventSenderTrust(mxEvent);
+        if (!eventSenderTrust) {
+            // We cannot find the device.  Instead, we have to verify the user.
+            const userTrust = await this.context.checkUserTrust(mxEvent.getSender());
+            this.setState({
+                verified: userTrust.isVerified() ? "user-verified": "warning",
```
Comment (foldleft) [inline src/components/views/rooms/EventTile.js]: actually, this logic is kinda wacky. probably not what we want?

File: src/components/views/rooms/EventTile.js [hunk 2]
```diff
@@ -282,18 +284,40 @@ export default createReactClass({
...
     _verifyEvent: async function(mxEvent) {
         if (!mxEvent.isEncrypted()) {
             return;
         }
 
+        // If we directly trust the device, short-circuit here
         const verified = await this.context.isEventSenderVerified(mxEvent);
+        if (verified) {
+            this.setState({
+                verified: "verified",
```
Comment (turt2live) [inline src/components/views/rooms/EventTile.js]: we might want to consider putting these string states into constants at some point.

File: src/components/views/rooms/EventTile.js [hunk 3]
```diff
@@ -282,18 +284,40 @@ export default createReactClass({
...
+                this.props.onHeightChanged();
+            });
+            return;
+        }
+
+        const eventSenderTrust = await this.context.checkEventSenderTrust(mxEvent);
+        if (!eventSenderTrust) {
+            this.setState({
+                verified: "unknown",
+            }, this.props.onHeightChanged); // Decryption may have cause a change in size
```
Comment (turt2live) [inline src/components/views/rooms/EventTile.js]: this isn't decryption though?

File: src/components/views/rooms/EventTile.js [hunk 4]
```diff
@@ -282,18 +284,40 @@ export default createReactClass({
...
+        }
+
         this.setState({
-            verified: verified,
-        }, () => {
-            // Decryption may have caused a change in size
-            this.props.onHeightChanged();
-        });
+            verified: eventSenderTrust.isVerified() ? "verified" : "warning",
+        }, this.props.onHeightChanged); // Decryption may have caused a change in size
```
Comment (turt2live) [inline src/components/views/rooms/EventTile.js]: likewise: this isn't decryption?

File: src/components/views/rooms/EventTile.js (hunk 3 above)
Comment (foldleft) [reply src/components/views/rooms/EventTile.js]: here I'm just copying from elsewhere in the function where I assume it was there for good reason. i think it's that if the element is decrypted then this causes different elements to render. I can take it out, if you like, but it migth be a regression.

File: src/components/views/rooms/EventTile.js [hunk 5]
```diff
@@ -66,6 +66,12 @@ const stateEventTileTypes = {
     'm.room.related_groups': 'messages.TextualEvent',
 };
 
+const E2ESTATE = {
```
Comment (turt2live) [inline src/components/views/rooms/EventTile.js]: this should probably be `E2E_STATE` but that's a silly reason to block this

**Pull request merged**